File: frontend/src/services/scenarioOrphans.ts
// =============================================================================
// scenarioOrphans.ts — client for the humane orphan strip (task 1.7C, D9)
// =============================================================================
//
// `GET /api/cases/:slug/scenarios/:scenarioId/facts/orphans`
//
// The saved references whose graph node no longer resolves, grouped by the
// document they came from. Measured on DEV 2026-08-03: 26 of them on scenario S-2,
// including six human INCLUDE decisions — the July 24/25 re-extraction loss,
// surfaced rather than hidden. They are the origin of the durability law.
//
// ## Why the grouping arrives already done
//
// A ref carries only a `graph_node_id`. The document it came from is encoded in
// that id (`doc-…:evidence:hash`), and the document's TITLE lives in Postgres —
// so parsing and labelling happen server-side, where both are reachable. A browser
// splitting id strings would be deriving provenance from an identifier, and it
// could not resolve a title at all (Standing Rule 12 and v2 §7 item 2).
//
// Same idioms as `scenarioFacts.ts`: `authFetch` (credentials + 30s
// AbortController timeout), `encodeURIComponent` on every path param,
// `readErrorMessage` to surface the backend `{message}`, and every non-2xx throws
// (Standing Rule 1 — an orphan strip that silently renders empty would say
// "nothing was lost", which is the one thing it must never say).

import { API_BASE_URL } from "./api";
import { authFetch } from "./auth";
import { readErrorMessage } from "./fetchUtils";

/**
 * How trustworthy a group's label is (backend `OrphanTitleState`).
 *
 * The state travels BESIDE the label because §2.8 forbids inventing titles, and a
 * single string cannot tell a real document title from a handle derived out of an
 * id. A client guessing from the shape of the text would guess wrong.
 */
export type OrphanTitleState = "resolved" | "unrecoverable" | "unparseable";

/** One document's worth of orphaned references (backend `OrphanGroup`). */
export type OrphanGroupDto = {
  /** The document id parsed from the refs' node ids. Empty for `unparseable`. */
  document_id: string;
  /** What to show: a real title, or an id-derived slug. Never invented. */
  label: string;
  title_state: OrphanTitleState;
  count: number;
};

/** The whole strip payload (backend `ScenarioOrphansResponse`). */
export type ScenarioOrphansDto = {
  /** Every orphaned reference, counted once — the summary line's number. Equals
   *  the sum of the groups' counts by construction, so a grouping failure cannot
   *  quietly shrink it. */
  total: number;
  /** Biggest loss first, then alphabetical. Empty when nothing is orphaned. */
  groups: OrphanGroupDto[];
};

/** Load this scenario's orphaned references, grouped by source document. */
export async function fetchScenarioOrphans(
  slug: string,
  scenarioId: string,
): Promise<ScenarioOrphansDto> {
  const response = await authFetch(
    `${API_BASE_URL}/api/cases/${encodeURIComponent(slug)}` +
      `/scenarios/${encodeURIComponent(scenarioId)}/facts/orphans`,
  );

  if (!response.ok) {
    const detail = await readErrorMessage(response);
    throw new Error(
      `Failed to load the unavailable references for scenario "${scenarioId}" ` +
        `(HTTP ${response.status}${detail}). These are saved rulings whose ` +
        `content is gone; the count matters, so this is not skipped silently.`,
    );
  }

  const data: unknown = await response.json();
  const parsed = data as Partial<ScenarioOrphansDto>;

  // Validate the load-bearing shape so a contract mismatch throws HERE with
  // context rather than as an `undefined.map` inside the strip.
  if (typeof parsed.total !== "number" || !Array.isArray(parsed.groups)) {
    throw new Error(
      `The unavailable-references response for scenario "${scenarioId}" is ` +
        `missing total/groups — backend/frontend contract mismatch. If this ` +
        `persists, report it to the site administrator.`,
    );
  }
  return parsed as ScenarioOrphansDto;
}

File: backend/src/dto/scenario_orphans.rs
//! Wire shapes for the humane orphan strip (task 1.7C, defect D9).
//!
//! `GET /cases/:slug/scenarios/:scenario_id/facts/orphans` — the saved references
//! whose graph node no longer resolves, grouped by the document they came from.
//!
//! A separate endpoint rather than a new field on the facts list, for two reasons:
//! design §2.8 moves the strip to the bottom of the page (it no longer hangs off
//! the ruling queue, so it no longer wants the queue's payload), and adding an
//! object wrapper to the existing bare-array response would break every caller for
//! no gain.

use serde::Serialize;

// RESPONSE-ONLY, so `Serialize` alone — the same shape as `dto::theme_scan`'s
// `ScanRunHeader`. Deriving `Deserialize` here would add a parse surface nothing
// uses: the backend never reads these back, and the tests construct them directly
// (which is what `PartialEq` is for). A `Deserialize` that no caller exercises is an
// unenforced contract, and the honest way to close that is to not have it rather
// than to bolt `deny_unknown_fields` onto a derive with no reader.

/// How trustworthy a group's label is.
///
/// ## Domain note: three different reasons a label is what it is
///
/// §2.8 forbids inventing titles, which means the client must be able to tell a
/// real document title from a handle derived out of an id. A single `label` string
/// cannot carry that distinction, and a client guessing from the shape of the text
/// would guess wrong. So the state travels beside the label — the same pairing as
/// `CardGrounding { state, label }`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum OrphanTitleState {
    /// The document still exists and this is its real title.
    Resolved,
    /// The document id parsed, but no such document exists any more — the label is
    /// a slug derived from the id. Measured: 2 of the live 26 refs, pointing at a
    /// `doc-court-of-appeals-ruling-…` id that predates the OCR re-titling to
    /// `…rulling…`.
    Unrecoverable,
    /// The node id did not carry a recognisable document prefix at all. Surfaced as
    /// its own group rather than dropped, so the total never disagrees with the
    /// database.
    Unparseable,
}

/// One document's worth of orphaned references.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct OrphanGroup {
    /// The document id parsed out of the refs' node ids. Empty for the
    /// `unparseable` group, which has no document to name.
    pub document_id: String,
    /// What the strip shows: the real title, or an id-derived slug. Never invented.
    pub label: String,
    /// Whether `label` is a real title — see [`OrphanTitleState`].
    pub title_state: OrphanTitleState,
    /// How many orphaned references came from this document.
    pub count: usize,
}

/// The whole orphan strip payload.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct ScenarioOrphansResponse {
    /// Every orphaned reference, counted once. This is the number the summary line
    /// says out loud ("26 saved references … content unavailable"), and it is the
    /// sum of the groups' counts BY CONSTRUCTION — including the unparseable
    /// group, so a grouping failure can never quietly shrink it.
    pub total: usize,
    /// Ordered biggest-loss-first, then alphabetically. Empty when nothing is
    /// orphaned — a legitimate, and very welcome, empty list.
    pub groups: Vec<OrphanGroup>,
}
